modules/misc/listenbrainz.c

```diff
@@ -0,0 +1,537 @@
+    add_string("listenbrainz_submission_url", "api.listenbrainz.org", URL_TEXT, URL_LONGTEXT, false)
+    set_capability("interface", 0)
+    set_callbacks(Open, Close)
+vlc_module_end ()
+
+static void DeleteSong (listen_t *p_song)
+{
+    FREENULL (p_song->psz_artist);
+    FREENULL (p_song->psz_album);
+    FREENULL (p_song->psz_title);
+    FREENULL (p_song->psz_musicbrainz_id);
+    FREENULL (p_song->psz_track_number);
+    p_song->date = 0;
+}
+
+static void ReadMetaData (intf_thread_t *p_this)
+{
+    intf_sys_t *p_sys = p_this->p_sys;
+
+    input_item_t *item = vlc_player_GetCurrentMedia (p_sys->player);
+    if ( item == NULL )
+        return;
+
+    //vlc_player_Lock(p_sys->player);
+    vlc_mutex_lock (&p_sys->lock);
+
+    p_sys->b_meta_read = true;
+    time(&p_sys->p_current_song.date);
+
+#define RETRIEVE_METADATA(a, b) do {                                    \
+        char *psz_data = input_item_Get##b(item);                       \
+        if (psz_data && *psz_data)                                      \
+            a = vlc_uri_encode(psz_data);                               \
+        else                                                            \
+            msg_Dbg (p_this, #b" missing.");                            \
+        free(psz_data);                                                 \
+    } while (0)
+
+    RETRIEVE_METADATA(p_sys->p_current_song.psz_artist, Artist);
+    if ( !p_sys->p_current_song.psz_artist )
+    {
+        DeleteSong (&p_sys->p_current_song);
+        goto error;
+    }
+
+    RETRIEVE_METADATA(p_sys->p_current_song.psz_title, Title);
+    if ( !p_sys->p_current_song.psz_title )
+    {
+        DeleteSong (&p_sys->p_current_song);
+        goto error;
+    }
+
+    RETRIEVE_METADATA(p_sys->p_current_song.psz_album, Album);
+    RETRIEVE_METADATA(p_sys->p_current_song.psz_musicbrainz_id, TrackID);
+    RETRIEVE_METADATA(p_sys->p_current_song.psz_track_number, TrackNum);
+    p_sys->p_current_song.i_length = SEC_FROM_VLC_TICK (input_item_GetDuration(item));
+    msg_Dbg (p_this, "Meta data registered");
+    vlc_cond_signal(&p_sys->wait);
+
+error:
+    vlc_mutex_unlock (&p_sys->lock);
+   // vlc_player_Unlock(p_sys->player);
+
+#undef RETRIEVE_METADATA
+}
+
+static listen_t CopySong (listen_t song)
+{
+    listen_t *copy = (listen_t *) malloc (sizeof (*copy));
+    memset (copy, 0, sizeof (*copy));
+    if ( song.psz_title )
+        copy->psz_title = strdup (song.psz_title);
+    if ( song.psz_artist )
+        copy->psz_artist = strdup (song.psz_artist);
+    if ( song.psz_album )
+        copy->psz_album = strdup (song.psz_album);
+    if ( song.psz_musicbrainz_id )
+        copy->psz_musicbrainz_id = strdup (song.psz_musicbrainz_id);
+    if ( song.psz_track_number )
+        copy->psz_track_number = strdup (song.psz_track_number);
+    copy->date = song.date;
+    return *copy;
+}
+
+static void Enqueue (intf_thread_t *p_this)
+{
+    bool b_skip = 0;
```

Comment:
I would prefer using goto error than this b_skip that make this function hard to read.